keps/sig-cluster-lifecycle/wgs/20200108-instance-specific-componentconfig.md

```diff
@@ -0,0 +1,418 @@
+
+The proposal is this: Add a new Kind for instance-specific config, and put
+instance-specific parameters in that object.
+
+This is relatively simple and isolates the parameters so that they can be
+dealt with separately. ComponentConfig files containing only sharable parameters
+can continue to be shared via a single ConfigMap. Files containing only
+instance-specific parameters can be provided to components via other means,
+such as the node startup scrips or an init container that inserts values from
+the Pod's Downward API.
+
+### Why not just keep using flags for instance-specific parameters?
+
+Flags have various problems, outlined in the [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c)
+doc.
+
+Leaving some parameters as flags while the rest are exposed via ComponentConfig
+results in an inconsistent API surface, and an inconsistent API versioning
+policy. Why fix only part of the problem when we can fix the whole thing?
+
+### Why not just auto-detect?
+
+One alternative that has been proposed in the past is to find a way to
+auto-detect values of instance-specific parameters and simply eliminate them
+from the configuration workflow. This is a great idea in theory, because it
+solves the problem by removing work for users, but it may be difficult in
+practice. For example, on a machine with multiple network cards, which IP should
+the kubelet use?
+
+Implementing the two-object approach does not prevent us from finding ways to
+auto-detect instance specific config in the future. It is worth noting that the
+_possibility_ of someday being able to auto-detect these has led to some
+hesitation to make a firm decision today. We have delayed for too long already,
+and a firm decision is needed to move forward.
+
+### Risks and Mitigations
+
+This creates an additional bucket for config parameters. We must be vigilant
+that only parameters that are truly instance-specific are added to the
+instance-specific Kind. We can provide warnings to developers via comments
+and tests to deter adding parameters to the instance-specific Kind that should
+instead be in the sharable Kind.
+
+For users deploying Pod templates, this solution can make the template more
+verbose. In the future, we can explore more elegant ways of generating files
+with instance-specific parameters correctly substituted. This proposal argues
+that the extra verbosity is worth the benefit of a consistent, versioned API for
+configuring core components.
+
+## Design Details
+
+ComponentConfig APIs are currently defined as a separate API group for each
+component, usually containing a single top-level Kind for configuring that
+component. The naming convention for the API group is `{component}.config.k8s.io`,
+and the convention for the Kind is `{Component}Configuration`. This KEP proposes
+that the naming convention for the new Kind be `{Component}InstanceConfiguration`.
```

Comment:
We can outline these as a set of best practices and not force them on components.
In the case of kubeadm, InitConfiguration and JoinConfiguration are instance specific kinds. They are different because of the specifics of the kubeadm workflow (InitConfiguration is used on `kubeadm init` which is done on a single machine - the first to be initialized to form a "cluster". JoinConfiguration is used on `kubeadm join` to join a machine to an existing cluster).

Most other components work as a service. They don't have user facing specific workflows as kubeadm. Hence a single instance specific kind is sufficient for them.
Not imposing, but recommending things here will make kubeadm compatible out-of-the-box with this KEP.

Another thing is the `{Component}` prefix. In my opinion it should be optional. For example, look at the current kube-proxy component config:
```yaml
apiVersion: kube-proxy.config.k8s.io/v1alpha1
kind: KubeProxyConfiguration
```

We have "Kube Proxy" in both the API group and the kind. I don't think that it sounds right to have `KubeProxyInstanceConfiguration` and `KubeProxyConfiguration` in the `kube-proxy.config.k8s.io` group. However, somebody could like it and want to stick to it.

In my opinion a viable option should be `InstanceConfiguration` and `SharedConfiguration` in `kube-proxy.config.k8s.io`. Some components may desire to use `ClusterConfiguration` in place of `SharedConfiguration` (as kubeadm does ATM).